File: tests/fixtures/python-library/library/services/catalog.py
from __future__ import annotations
from typing import Generic, TypeVar
from library.interfaces.searchable import Searchable

T = TypeVar("T", bound=Searchable)


class Catalog(Generic[T]):
    """A catalog that holds searchable items."""

    class Stats:
        """Nested class: statistics about the catalog."""

        def __init__(self, total_items: int, name: str) -> None:
            self.total_items = total_items
            self.name = name

    def __init__(self, name: str) -> None:
        self._items: list[T] = []
        self._name = name

    def add(self, item: T) -> None:
        """Add an item to the catalog."""
        self._items.append(item)

    def search(self, query: str) -> list[T]:
        """Search for items matching a query."""
        return [item for item in self._items if query in item.search_text()]

    def stats(self) -> Stats:
        """Get catalog statistics."""
        return self.Stats(total_items=len(self._items), name=self._name)


def create_default_catalog() -> Catalog:
    """Free function: create a default catalog for books."""
    return Catalog(name="Main Library")

File: tests/fixtures/rust-library/src/services/catalog.rs
use crate::traits::searchable::Searchable;

/// A catalog that holds searchable items.
pub struct Catalog<T: Searchable> {
    items: Vec<T>,
    name: String,
}

/// Nested type: statistics about the catalog.
pub struct CatalogStats {
    pub total_items: usize,
    pub name: String,
}

impl<T: Searchable> Catalog<T> {
    /// Create a new empty catalog.
    pub fn new(name: String) -> Self {
        Self {
            items: Vec::new(),
            name,
        }
    }

    /// Add an item to the catalog.
    pub fn add(&mut self, item: T) {
        self.items.push(item);
    }

    /// Search for items matching a query.
    pub fn search(&self, query: &str) -> Vec<&T> {
        self.items
            .iter()
            .filter(|item| item.search_text().contains(query))
            .collect()
    }

    /// Get catalog statistics.
    pub fn stats(&self) -> CatalogStats {
        CatalogStats {
            total_items: self.items.len(),
            name: self.name.clone(),
        }
    }
}

/// Free function: create a default catalog for books.
pub fn create_default_catalog() -> Catalog<crate::models::book::Book> {
    Catalog::new("Main Library".to_string())
}
